core/ipc/twin/common/src/main/java/org/opennms/core/ipc/twin/common/AbstractTwinSubscriber.java

```diff
@@ -230,72 +136,183 @@ protected TwinUpdate mapTwinResponseToProto(byte[] responseBytes) {
+            this.key = Objects.requireNonNull(key);
         }
 
-        public void accept(byte[] objValue) throws IOException {
-            final T value = objectMapper.readValue(objValue, clazz);
-            LOG.trace("Updated consumer with key {}", key);
-            consumer.accept(value);
+        /**
+         * Consume the subscription.
+         *
+         * Adds a consumer to the subscription. Incoming values will be forwarded to the passed consumer until the
+         * returned value is closed. If there is a value already available for the subscription, the consumer will be
+         * called with this value immediately.
+         *
+         * @param clazz The class of the value to consume
+         * @param consumer The consumer accepting the values
+         * @param <T> The class of the value to consume
+         * @return a Closable, stopping the consumption when closed
+         */
+        public synchronized <T> Closeable consume(final Class<T> clazz, final Consumer<T> consumer) {
+            final Consumer<JsonNode> jsonConsumer = (json) -> {
+                try {
+                    // Deserialize to the final class
+                    final var value = AbstractTwinSubscriber.this.objectMapper.treeToValue(json, clazz);
+
+                    // Forward to typed consumer
+                    consumer.accept(value);
+
+                } catch (final JsonProcessingException e) {
+                    LOG.error("Deserialize twin failed: {} as {}", this.key, clazz, e);
+                }
+            };
+
+            if (this.value == null) {
+                // Initially request value
+                this.request();
+            } else {
+                // If value already exists, forward to consumer without requesting
+                jsonConsumer.accept(this.value.value);
+            }
+
+            // Add the consumer to the subscription
+            this.consumers.add(jsonConsumer);
+
+            // Return the closable removing the consumer
+            return () -> {
+                this.consumers.remove(jsonConsumer);
+
+                // Drop the subscription if it becomes empty
+                if (this.consumers.isEmpty()) {
+                    AbstractTwinSubscriber.this.subscriptions.remove(this.key);
+                }
+            };
         }
 
-        @Override
-        public boolean equals(Object o) {
-            if (this == o) return true;
-            if (o == null || getClass() != o.getClass()) return false;
-            SessionImpl<?> session = (SessionImpl<?>) o;
-            return Objects.equals(key, session.key) && Objects.equals(consumer, session.consumer) && Objects.equals(clazz, session.clazz);
+        private synchronized void accept(final Value value) {
+            Objects.requireNonNull(value);
+
+            // Skip if value has not changed
+            if (this.value != null && Objects.equals(this.value.value, value.value)) {
+                return;
+            }
```

Comment:
What happens if the value does not change, but the session/version? I think in that case the `Value` should be replaced but consumers need not to be notified.

Reply:
Why should the value be replaced if it does not change? The equal-semantics on a JSON object are well-defined. There is no harm in keeping the old object.

Reply:
The session/version could have changed.

Reply:
granted.

Reply:
AFAIS it is not granted. `accept` is called if the sessionId or version has changed and the update was not a patch.